formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt

```diff
@@ -109,86 +106,109 @@ internal open class StreamingJsonDecoder internal constructor(
             }
         }
+        if (hasComma) reader.fail("Unexpected trailing comma")
         return CompositeDecoder.DECODE_DONE
     }
 
-    private fun decodeListIndex(tokenClass: Byte): Int {
-        // Prohibit leading comma
-        if (tokenClass != TC_COMMA && currentIndex != -1) {
-            reader.requireTokenClass(TC_END_LIST) { "Expected end of the array or comma" }
-        }
-        return if (!reader.canBeginValue) {
-            reader.require(tokenClass != TC_COMMA) { "Unexpected trailing comma" }
-            CompositeDecoder.DECODE_DONE
+    private fun handleUnknown(key: String): Boolean {
+        if (configuration.ignoreUnknownKeys) {
+            reader.skipElement()
         } else {
+            reader.failOnUnknownKey(key)
+        }
+        return reader.tryConsumeComma()
+    }
+
+    private fun decodeListIndex(): Int {
+        // Prohibit leading comma
+        val hasComma = reader.tryConsumeComma()
+        return if (reader.canConsumeValue()) {
+            if (currentIndex != -1 && !hasComma) reader.fail("Expected end of the array or comma")
             ++currentIndex
+        } else {
+            if (hasComma) reader.fail("Unexpected trailing comma")
+            CompositeDecoder.DECODE_DONE
         }
     }
 
+
     override fun decodeBoolean(): Boolean {
         /*
          * We prohibit non true/false boolean literals at all as it is considered way too error-prone,
          * but allow quoted literal in relaxed mode for booleans.
          */
-        val string = if (configuration.isLenient) {
-            reader.takeString()
+        return if (configuration.isLenient) {
+            reader.consumeBooleanLenient()
         } else {
-            reader.takeBooleanStringUnquoted()
+            reader.consumeBoolean()
         }
-        string.toBooleanStrictOrNull()?.let { return it }
-        reader.fail("Failed to parse type 'boolean' for input '$string'")
     }
 
     /*
-     * The rest of the primitives are allowed to be quoted and unqouted
+     * The rest of the primitives are allowed to be quoted and unquoted
      * to simplify integrations with third-party API.
      */
-    override fun decodeByte(): Byte = reader.parseString("byte") { toByte() }
-    override fun decodeShort(): Short = reader.parseString("short") { toShort() }
-    override fun decodeInt(): Int = reader.parseString("int") { toInt() }
-    override fun decodeLong(): Long = reader.parseString("long") { toLong() }
+    override fun decodeByte(): Byte {
+        val value = reader.consumeNumericLiteral()
+        // Check for overflow
+        if (value != value.toByte().toLong()) reader.fail("Failed to parse byte for input '$value'")
+        return value.toByte()
+    }
+
+    override fun decodeShort(): Short {
+        val value = reader.consumeNumericLiteral()
+        // Check for overflow
+        if (value != value.toShort().toLong()) reader.fail("Failed to parse byte for input '$value'")
```

Comment:
Should the error message be adjusted (same for `decodeInt`)?
```suggestion
        if (value != value.toShort().toLong()) reader.fail("Failed to parse short for input '$value'")
```

Reply:
Thanks!